### DISCUSSION TITLE
Just for my curiosity: roboto-fontface version resolve

### DISCUSSION THREAD

**Discussion by jorismak** [pnpm/pnpm#6333] [Q&A]
Hi,

Just started doing a little test on a project using pnpm.

And I see something that makes me go "eh??!".

In my package.json file, is listed `roboto-fontface` with version constraint `*`.

```json
{
    "dependencies": {
        "roboto-fontface": "*"
    }
}
```

`npm` and `yarn` resolve this to version `0.10.0`.

But if I start over clean, and do a `pnpm install`, it resolves `roboto-fontface` to version `0.4.1`.

If I do a `pnpm add roboto-fontface`, it still leaves it at 0.4.1.
But it does change the constraint in package.json:

```
        "roboto-fontface": "^0.4.1",
````

Now, if I tell it to really get 0.10:
`pnpm add roboto-fontface@^0.10` I 'listens' correctly:

```
Packages: +1 -1
+-
Progress: resolved 128, reused 106, downloaded 0, added 0, done

dependencies:
- roboto-fontface 0.4.1
+ roboto-fontface 0.10.0

Done in 1.5s
```

So, why does pnpm seem to think that 0.4.1 is the latest available, while other package managers take 0.10.0?
pnpm clearly seems to know of version 0.10.0, otherwise it couldn't install it :).

**Comment by RexSkz** [comment]
On my computer it seems correct:

```text
$ pnpm add roboto-fontface
Packages: +1
+
Packages are hard linked from the content-addressable store to the virtual store.
  Content-addressable store is at: /Users/rexzeng/Library/pnpm/store/v3
  Virtual store is at:             node_modules/.pnpm

dependencies:
+ roboto-fontface 0.10.0

Progress: resolved 1, reused 1, downloaded 0, added 1, done
Done in 1.8s
```

**Comment by jorismak** [comment]
It's a Vue2 / Vuetify2 project. Maybe one of the dependencies has the font
listed as a (peer) dependency, influencing the version picked.

But still, npm and yarn pick 0.10.0 without duplicates.

I just tried in an empty directory, with a fresh new `pnpm init`
package.json file. And yes, there it takes 0.10.0.

I just tried another random Vue2 project (with npm), threw away the
package-lock.json file and node_modules file, did another `pnpm install`
and it gave me 0.4.1 again.
I'm 'grepping' through the entire node_modules folder (with -R to search in
symlinks) and I can't find anything that might cause this.

`pnpm outdated` shows that 0.10.0 is available (so it sees it as higher).
`pnpm why roboto-fontface`  lists only my own package / project.

Actually running `pnpm up` does upgrade the package from 0.4.1 to 0.10.0.
So apparently there is nothing 'holding it back'. But still a `pnpm
install` adds an older version on very first run. I'm still confused :).

**Comment by zkochan** [comment]
It is because pnpm v8 now has [resolution-mode](https://pnpm.io/npmrc#resolution-mode) set to `lowest-direct` by default. It means that direct dependencies are resolved to the lowest versions that satisfy the range.

Overall, it is never a good idea to add a package with the `*` range.

Resolving direct dependencies to the lowest version is a good approach because it guarantees that you test your package with the dependency with which it might get installed. In most cases it will be installed with the latest version satisfying the range. However, let's assume your dependency is `foo@^1.0.0`, the latest version of foo is `1.1.0` and the project installs your package has a direct dependency to `foo@1.0.0`. In this case, `foo` will be deduplicated and `foo` v1.0.0 will be installed as a dependency of your package. So, if during local development you were also using v1.0.0, then you can be sure that you did not use any features that were added only in v1.1.0.

**Comment by jorismak** [comment]
Ok.. if the default is set to lowest-direect by default then it makes sense.

Over the how and why, I'm a bit baffled, but that's another discussion and
not for here. The fact that other package managers have a different default
speaks that the opinions about this are divided online. And the lockfile is
there to prevent version differences across different installs, so this
shouldn't happen in practise.

I'm off looking how to change the default, thanks!
